clippy_lints/src/field_reassign_with_default.rs

```diff
@@ -0,0 +1,208 @@
+use crate::utils::{match_def_path, paths, qpath_res, snippet, span_lint_and_note};
+use if_chain::if_chain;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::Res;
+use rustc_hir::{Block, Expr, ExprKind, PatKind, QPath, Stmt, StmtKind};
+use rustc_middle::ty::{self, Adt, TyS};
+use rustc_span::symbol::{Ident, Symbol};
+
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for immediate reassignment of fields initialized
+    /// with Default::default().
+    ///
+    /// **Why is this bad?**It's more idiomatic to use the [functional update syntax](https://doc.rust-lang.org/reference/expressions/struct-expr.html#functional-update-syntax).
+    ///
+    /// **Known problems:** Assignments to patterns that are of tuple type are not linted.
+    ///
+    /// **Example:**
+    /// Bad:
+    /// ```
+    /// # struct A { i: i32 }
+    /// let mut a: A = Default::default();
+    /// a.i = 42;
+    /// ```
+    /// Use instead:
+    /// ```
+    /// # struct A { i: i32 }
+    /// let a = A {
+    ///     i: 42,
+    ///     .. Default::default()
+    /// };
+    /// ```
+    pub FIELD_REASSIGN_WITH_DEFAULT,
+    style,
+    "binding initialized with Default should have its fields set in the initializer"
+}
+
+declare_lint_pass!(FieldReassignWithDefault => [FIELD_REASSIGN_WITH_DEFAULT]);
+
+impl LateLintPass<'_> for FieldReassignWithDefault {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &Block<'_>) {
+        // find all binding statements like `let mut _ = T::default()` where `T::default()` is the
+        // `default` method of the `Default` trait. and store statement index in current block being
+        // checked and the name of the bound variable
+        let binding_statements_using_default: Vec<(usize, Symbol, &TyS<'_>)> =
```

Comment:
```suggestion
        let binding_statements_using_default =
```
Although maybe you added the annotation for clarity?